### PR TITLE
feat: ng environment commands

### PR DESCRIPTION
**Pull request by ysndr**
This PR implements environment commands for local environments (`.flox` based), replacing the existing abstraction of project environments.

Commands implemented:

* `flox init` (including naming, renamed "old" init -> `package-init`) 
* `flox delete`
* `flox install`
* `flox uninstall`
* `flox edit`
* `flox list` - list packages
* `flox envs` - list environments
* `flox activate` - nix shell based implementation to be iterated on

Remaining local commands:

* `flox upgrade`

### CODE REVIEW

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment.rs:115]:
```suggestion
        debug!("building with nix ....");
```

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment.rs:489]: I believe we plan to support only one environment per directory. I'm not opposed to this implementation, but I think we should be careful to not waste time future proofing. There's a few things throughout that we should change throughout if we're going to enforce a single environment.

But I don't think a `.flox` directory should have any concept of owner (unless it's a link to FLOX_META, but I think that's something different)

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment.rs:536]:
```suggestion
    pub async fn init(
```

File: crates/flox-rust-sdk/src/models/environment_ref.rs
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment_ref.rs:1]: I think we'll still need the logic to handle defaulting named environments for `-e`, but I don't think we'll support multiple environments in a `.flox`. So I'm not sure we want to drop all the named handling, and I'm also not sure the way this implements `.flox` handling is correct

File: crates/flox-rust-sdk/src/models/flox_package.rs
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/flox_package.rs:57]: Looks like you're unwrapping this

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment.rs:124]: Is this a TODO? We'll need to create a link, right?

**Review by mkenigs — COMMENTED** [review]
Looks like a lot of progress! Left a few comments but here are my higher level thoughts:
- I'm not sure this is quite the handling we want for owner/name and project vs named environments. I think that behavior may be fully specced out as of today. I wasn't thinking we need to handle all of that correctly to work on the project env implementations. I would either: 1) tackle the environment naming stuff on its own or 2) just do the project commands and try to avoid touching the environment naming stuff (i.e. don't drop all our floxmeta resolution logic)
- This is too big of a change. I think for the sake of speed we'll either need to just merge what you have, and improve on it (maybe putting back in a few of the things that got deleted and making sure we don't break too much). Or we need to split out the skeletal work, review and merge that, and then work on commands one at a time. But a 1500 line change is just going to block us if we actually try to review it

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment.rs:489]: This work was roughly based on the project envs we used to have, thus .flox here can hold multiple environments (or links, but i digress).
The spec mentions "needs engineering" and i remember we were discussing that point. however, a single env per directory would make the `DotFloxDir` struct either simpler or close to unnecessary (pending link handling...)

File: crates/flox-rust-sdk/src/models/environment_ref.rs
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment_ref.rs:1]: My understanding was that we drop the "named" envs, but that .flox can either be a local env with its definitions in place, or a remote env with its staged definition managed centrally (within a floxmeta clone) subject to sync with floxhub.

Thus you dont access "remote" envs directly, but by way of linking it to your current dir (+ .flox).
That kinda makes more sense when we'd allow multiple envs per .flox, but i guess that's a good point to circle around.

**Comment by ysndr** [comment]
> This is too big of a change.

This PR is certainly quite sizeable, and more of a demonstration that we can start iterating on something, within a fairly short amount of time.
I don't see this as the prime example of adding changes to this repo but rather a jump start from which to start working on.
Alternatively, I'm also happy to close this PR and slice it up into individual parts reviewed (mostly) one at a time to chew through the review of this phase.

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (ysndr) [inline crates/flox-rust-sdk/src/models/environment.rs:1]: If you could have a look through the _structure_ of this.
I believe the precise implementation of the commands will change soon enough but the interface should remain roughly the same.

General "getters" are implemented for `Environment<_>`,

methods on the "atomic" environment in the .flox dir (mainly constructors) are implemented for  `Environment<Read>`.

You create a temporary writable copy of the environment with `Environment<Read>::modify_in`

Writing commands are implemented on `Environment<Modify>`.


Does that make sense?

I guess as the transaction rules are different for "remote" environments there will be a common trait around there but I wouldn't go into too much detail on that right now.

File: crates/flox-rust-sdk/src/models/flox_package.rs
Comment (ysndr) [inline crates/flox-rust-sdk/src/models/flox_package.rs:1]: This parses _the current_ definition of "installable" packages.
The parsing rules for this likely change as the spec progresses.

`flox_nix_attribute` is an impl detail of the current flox.nix, it produces attrpaths to be created (or deleted) from the nix source (using the rnix utility)

Don't spent too much time looking over these, they will change with the new env backend

File: crates/flox-rust-sdk/src/utils/rnix.rs
Comment (ysndr) [inline crates/flox-rust-sdk/src/utils/rnix.rs:1]: Manipulating ASTs is fun...
We should review this eventually as there seems to be no proper substitute, except not using nix (or dumber nix)

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (zmitchell) [inline crates/flox-rust-sdk/src/models/environment.rs:44]: I think we should clarify what it means to "modify" the environment. Do we only include modifications to the declarative definition of the environment, or do we also include operations that have side effects e.g. `build`? For example, `build` doesn't change the environment definition, but it has side effects in that it overwrites `result`.

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (zmitchell) [inline crates/flox-rust-sdk/src/models/environment.rs:137]: I'm confused, I thought the catalog kept track of which packages were built or provided by an environment. This says "installed from the environment". Is this method returning a list of packages _built from this environment_ that are also _installed in this environment_?

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (zmitchell) [inline crates/flox-rust-sdk/src/models/environment.rs:168]: Again, when you say "in the environment," do you mean installed into the environment, or the packages provided by the environment?

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (zmitchell) [inline crates/flox-rust-sdk/src/models/environment.rs:179]: What's the intended use case for this?

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (zmitchell) [inline crates/flox-rust-sdk/src/models/environment.rs:315]: I think replacing the file content and verifying the build are two separate operations that shouldn't be mixed together in the same low-level method. I recommend separate low-level `replace_definition` and `try_build` methods even though they'll be very small methods. This makes testing easier because you can mock out individual operations that either do I/O or aren't what you're concerned with. When in doubt I prefer to provide building blocks and defer to the caller to put them together how they need to (the `flox edit` command handler in this case).

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (zmitchell) [inline crates/flox-rust-sdk/src/models/environment.rs:1]: I've said this several times at this point, but I think `Environment`, `EnvironmentRead`, and `EnvironmentModify` should be traits implemented on some struct. I think the `Environment` struct is fine for this. Then you can have `impl EnvironmentRead for Environment<Read>` and `impl EnvironmentModify for Environment<Modify>`. This will make testing much more flexible and more clearly separate the `trait Environment` interface from the `struct Environment` internals.

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (zmitchell) [inline crates/flox-rust-sdk/src/models/environment.rs:434]: I'm not sure this should be the responsibility of the `Environment`. I would expect this to be done by the `flox edit` command. I see the `Environment` as mostly a container, and it feels like a boundary violation to allow the container to overwrite the origin.

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/environment.rs:537]: Agreed, we should find ways to drop `async` where possible

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (tomberek) [inline crates/flox-rust-sdk/src/models/environment.rs:40]: This seems to limit this to only "path://" usages. This means all other machinery to talk about remote resources using URLs will be blocked.

This can be a TODO for later, but it seems to be an artificial limitation when we already have a language to talk about a wide range of non-local envs.

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (tomberek) [inline crates/flox-rust-sdk/src/models/environment.rs:102]: Or use the top-level flox.nix?

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (tomberek) [reply crates/flox-rust-sdk/src/models/environment.rs:120]: We used to put these into a .flox ... can put them into a .flox/cache?

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (tomberek) [reply crates/flox-rust-sdk/src/models/environment.rs:137]: This seems to be performing an eval to determine what catalog entries there are in the flake.

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (mkenigs) [reply crates/flox-rust-sdk/src/models/environment.rs:102]: The current template uses pkgs/default so just using that is less invasive. I expect this will change when we revisit mkFloxEnv

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (mkenigs) [reply crates/flox-rust-sdk/src/models/environment.rs:40]: We've deliberately restricted the scope of this first iteration to make it simpler. But yes we can revisit

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (mkenigs) [reply crates/flox-rust-sdk/src/models/environment.rs:434]: Would you want to move this method to `Environment<Read>`? If we do so it would probably make more sense to name them `Environment<Original>` and `Environment<Temporary>`?

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (mkenigs) [reply crates/flox-rust-sdk/src/models/environment.rs:1]: Makes sense to me

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (mkenigs) [reply crates/flox-rust-sdk/src/models/environment.rs:434]: I also think in some cases it would make more sense to have methods like install on `Environment<Original>` even though they modify the environment. E.g. it would be easier for flox CLI to consume `install` if it didn't have to worry about calling `modify_in`. In `Environment<Original>::install` we could have:
```
temporary = modify_in()
temporary.set_environment() // we could even have set_environment container the modify_in call
self.finish(temporary)
```

That simplifies the API (just call install, rather than calling modify_in and install), and I don't think we lose anything in terms of transactionality

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/environment.rs:137]: Then I would advocate for a very explicit name like `catalog_entries_provided` or something similar. We aren't paying per character and we should strive to make it impossible to misunderstand what the method is doing.

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (zmitchell) [reply crates/flox-rust-sdk/src/models/environment.rs:434]: I think it makes less sense for `install` to be on `Environment<Read>` because I read the `State` as what permissions you have, and it doesn't make sense to me to be able to install something when I have read-only permissions.

I like the `Original`/`Temporary` distinction. In that case I would see a flow going like this:
```rust
fn try_install(self: Environment<Original>, pkg: Package) -> Result<(), SomeError> {
    let mut temp_env: Environment<Temporary> = self.make_temporary()?;
    let new_contents = temp_env.contents_with_new_package(pkg)?;
    temp_env.update_file(new_contents);
    let succeeded = temp_env.build()?;
    if succeeded {
        self.replace_with(temp_env)?;
        return Ok(());
    } else {
        temp_env.clean_up()?;
        bail!("could not install package")
    }
}
```

Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment.rs:434]: Agree a split `Original`/`Temporary` and only consume methods from `Original` seem useful.

Some concerns:

* should `flox install a b c` (b being invalid) a) fail atomically b) install `a` c) install `a` & `c`? 
  `try_install` taking a single package we can't provide a). In remote envs (considering this might be a trait) we'd also generate multiple generations (one per call to try_install)
  I dont think we have to block on any of that now, seems simple enough to change.
  
* We still need to pass in the tempdir at some point, either store it in `Original` or pass as argument

* You prefer methods like `contents_with_new_package` that would just give you an updated str rather than editing the (temporary) rather than `contents_with_new_package` calling `update_file` itself?

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment.rs:1]: Agree we likely want these within a trait eventually.

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment.rs:315]: Yeah the building within set environment is a it odd, providing that as a separate method or part of `finish` would make more sense (i.e. you shouldn't `finish`/confirm the "transaction" without verifying it can build).
However with the new decalrative spec the build part will be substantially more complex.

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment.rs:168]: Are those not the same thing?
The environment provides the packages that are installed to it, anything that might come with it by means of `propagatedBuildInputs` and such is not a guaranteed

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment.rs:179]: Garbage collection of environments
We can disregard that for now i think (doesn't apply to local envs anyway).

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment.rs:120]: The moving had some issues with the symlink so i commented this out.
We will need this just deferred to figuring out "moving directories"

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment.rs:434]: To separate things a bit more I previously thought of using A pattern closer to `Publish`

* a trait for environments exposing _readable_ fields (path, installable, owner, name, packages, etc, build(?) )

* a struct for the impl of modify envs :

``` rust
struct ModifyEnv<E> {
	original: E,
    temporary: LocalEnv
}

impl <E: Environment> ModifyEnv<E> {
    fn env_declaration_with_new_package(&self) -> Result<String>; 
    fn env_declaration_without_package(&self) -> Result<String>;

	fn set_env_declaration_content(&mut self, String) -> Result<()>; 
    fn try_build(&self) -> Result<()>;
}

// apply changes, differently per "Environment backend"

impl ModifyEnv<LocalEnv> {
    fn apply(self) -> Result<LocalEnv>;
}


impl ModdifyEnv<RemoteEnv> {
    fn apply(self) -> Result<RemoteEnv>;
}
```

* A struct to impl activate 
* A struct for push/pull operations


this way we should be able to work on separate subsystems without having to touch the same struct or trait that much.
This would also make it easier to implement some parts in other languages through json interfaces without pulling all that into the environment itself.

**Comment by zmitchell** [comment]
I'd just like to point out that even after #213 there's still work to do with splitting the trait into the methods we discussed yesterday. I'm posting them here for visibility:
```
/// Generic environment implementation
pub trait Environment {
    pub fn install(pkgs: Packages) -> Result<(), Error>;
    pub fn make_temporary(&self) -> impl TemporaryEnvironment;
    pub fn build();
    pub fn contents_with_new_package() -> String;
    pub fn contents_with_package_removed() -> String;
    pub fn replace_with(new_env: impl Environment);
    pub fn owner();
    pub fn name();
    pub fn get_installable();
    pub fn push();
    pub fn pull() -> Result<(), Error>;
    pub fn activate();
}

pub trait TemporaryEnvironment: Environment {
    fn update_file(contents: String);
}
```

I think it's important to keep granular methods available even if they're just tied together in higher level methods like `install` that are meant for consumers.

**Comment by zmitchell** [comment]
@ysndr and I discussed and we think it's probably better to leave the transactional machinery out of the trait since different contexts will require transactions to take different forms e.g. a local environment just needs to edit the environment definition, but you might want to push a remote environment as part of the transaction. If you include the transaction machinery in the trait you end up with a trait that contains the union of all of those different transaction methods, and then `PathEnvironment` needs to include implementations of transaction methods that it may not even use.

We also agree that it makes sense to remove any validation or editing of the environment definition from the trait since that's not _really_ part of what defines the interface of an environment. All of that functionality can be implemented as free-standing functions, which is how it's currently written in the PR.

The "original" vs. "temporary" distinction is no longer part of the trait definition since that distinction is only relevant in the context of transactions, which as I said above are no longer part of the trait.

So, we sketched out something that looks like this:
```rust
pub trait Environment {
    // High level environment operations
    fn install();
    fn uninstall();
    fn edit();
    fn build();
    fn activate();
    // Getters for metadata that all environments have
    fn owner();
    fn name();
}

// A particular environment type
pub struct SomeEnv<S> {};

impl SomeEnv<S> {
    // All of the transaction machinery
    fn make_temporary() {}
    fn update_file() {}
}

// Functions for editing environment defintions
fn add_packages(current_contents: String) -> Result<String, Error> {}
fn remove_packages(current_contents: String) -> Result<String, Error> {}
```

**Comment by zmitchell** [comment]
Discussed with @mkenigs, I'll pick this up and make the required changes so we can try to get this reviewed and merged quickly

**Comment by mkenigs** [comment]
Whoops just realized I've been leaving comments on https://github.com/flox/flox/commit/88f147c1783e0b0eefcc7d7a20ab4691a2c68c0c

**Review by mkenigs — APPROVED** [review]
Assuming threads on https://github.com/flox/flox/commit/88f147c1783e0b0eefcc7d7a20ab4691a2c68c0c get resolved LGTM

**Comment by zmitchell** [comment]
Those comments should be taken care of. @ysndr I'll let you do the honors of the massive rebase and the merge. I started it and then lost confidence that I knew which changes should make it into the final merge.

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (ysndr) [inline crates/flox-rust-sdk/src/models/environment.rs:195]: So i still don't see why `PathEnvironment<Temporary>` has to comply to this interface..
Yes, we want to build and use some methods we share with `PathEnvironment<Original>`, but why not implment the subset of methods as private methods for `PathEnvironment<_>` and call those in the `Environment` implementation for `PathEnvironment<Temporary>`?

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (ysndr) [inline crates/flox-rust-sdk/src/models/environment.rs:498]:
```suggestion
/// Created by [`PathEnvironment<Original>::make_temporary`].
```

File: crates/flox-rust-sdk/src/models/environment.rs
Comment (ysndr) [inline crates/flox-rust-sdk/src/models/environment.rs:102]: Does this really need to be a trait? I thought we want to keep the whole temporary/transaction part out of the interface.

**Review by ysndr — COMMENTED** [review]
Non blocking comments on the geography, let's get this in and figure out the rest as we go (incl, `edit`).

**Pull request merged**
